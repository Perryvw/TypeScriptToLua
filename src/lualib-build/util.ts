import * as tstl from "..";

export function isExportTableDeclaration(node: tstl.Node): node is tstl.VariableDeclarationStatement & { left: [] } {
    return tstl.isVariableDeclarationStatement(node) && isExportTable(node.left[0]);
}

export function isExportTable(node: tstl.Node): node is tstl.Identifier {
    return tstl.isIdentifier(node) && node.text === "____exports";
}

export type ExportTableIndex = tstl.TableIndexExpression & { index: tstl.StringLiteral };
export function isExportTableIndex(node: tstl.Node): node is ExportTableIndex {
    return tstl.isTableIndexExpression(node) && isExportTable(node.table) && tstl.isStringLiteral(node.index);
}

export function isExportAlias(
    node: tstl.Node,
): node is tstl.VariableDeclarationStatement & { right: [ExportTableIndex] } {
    return tstl.isVariableDeclarationStatement(node) && node.right !== undefined && isExportTableIndex(node.right[0]);
}

export type ExportAssignment = tstl.AssignmentStatement & { left: [ExportTableIndex] };
export function isExportAssignment(node: tstl.Node): node is ExportAssignment {
    return tstl.isAssignmentStatement(node) && isExportTableIndex(node.left[0]);
}

export function isRequire(node: tstl.Node) {
    return (
        tstl.isVariableDeclarationStatement(node) &&
        node.right &&
        tstl.isCallExpression(node.right[0]) &&
        tstl.isIdentifier(node.right[0].expression) &&
        node.right[0].expression.text === "require"
    );
}

export function isImport(node: tstl.Node, importNames: Set<string>) {
    return tstl.isVariableDeclarationStatement(node) && importNames.has(node.left[0].text);
}

export function isExportsReturn(node: tstl.Node) {
    return (
        tstl.isReturnStatement(node) &&
        tstl.isIdentifier(node.expressions[0]) &&
        node.expressions[0].text === "____exports"
    );
}
