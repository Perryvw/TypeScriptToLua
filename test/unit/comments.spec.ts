import * as util from "../util";

test("JSDoc is copied on a function", () => {
    const builder = util.testModule`
        /**
         * This is a function comment.
         * It has multiple lines.
         */
        function foo() {}
    `.expectToHaveNoDiagnostics();

    const transpiledFile = builder.getLuaResult().transpiledFiles[0];
    expect(transpiledFile).toBeDefined();
    const { lua } = transpiledFile;
    expect(lua).toBeDefined();
    expect(lua).toContain("This is a function comment.");
});

test("JSDoc is copied on a variable", () => {
    const builder = util.testModule`
        /**
         * This is a variable comment.
         * It has multiple lines.
         */
        const foo = 123;
    `.expectToHaveNoDiagnostics();

    const transpiledFile = builder.getLuaResult().transpiledFiles[0];
    expect(transpiledFile).toBeDefined();
    const { lua } = transpiledFile;
    expect(lua).toBeDefined();
    expect(lua).toContain("This is a variable comment.");
});
