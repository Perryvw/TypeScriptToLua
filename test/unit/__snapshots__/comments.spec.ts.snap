// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSDoc is copied on a function: code 1`] = `
"-- This is a function comment.
-- It has multiple lines.
function foo(self)
end"
`;

exports[`JSDoc is copied on a function: diagnostics 1`] = `""`;

exports[`JSDoc is copied on a variable: code 1`] = `
"-- This is a variable comment.
-- It has multiple lines.
foo = 123"
`;

exports[`JSDoc is copied on a variable: diagnostics 1`] = `""`;
